### PR TITLE
Fixes HTTP2 removal for GCS

### PR DESCRIPTION
**Pull request by cyriltovena**
This PR fixes a bug in the removal of http2 causing corrupted TLS handshake. The bug seems to be when disabling http by setting `transport.TLSNextProto` to a non nil map on a cloned transport. Looking at what the clone function does it seems to initialize and enforce TLS for http2, which mixes http2 TLS with http1 requests. This seems worth adding a issue on the go repository.

I fixes the bug by not cloning the transport but copy pasting the default one.

The PR also allow to remove http2 only for GETs, since we don't need this for other method.

### CODE REVIEW

**Comment by dannykopping** [comment]
> This PR fixes a bug in the removal of http2 causing corrupted TLS handshake. The bug seems to be when disabling http by setting `transport.TLSNextProto` to a non nil map on a cloned transport. Looking at what the clone function does it seems to initialize and enforce TLS for http2, which mixes http2 TLS with http1 requests. This seems worth adding a issue on the go repository.
> 
> I fixes the bug by not cloning the transport but copy pasting the default one.
> 
> The PR also allow to remove http2 only for GETs, since we don't need this for other method.

We should definitely open that issue from the sound of this

File: pkg/storage/chunk/gcp/gcs_object_client.go [hunk 1]
```diff
@@ -84,7 +84,12 @@ func newGCSObjectClient(ctx context.Context, cfg GCSConfig, hedgingCfg hedging.C
 
 func newBucketHandle(ctx context.Context, cfg GCSConfig, hedgingCfg hedging.Config, hedging bool, clientFactory ClientFactory) (*storage.BucketHandle, error) {
 	var opts []option.ClientOption
-	httpClient, err := gcsInstrumentation(ctx, storage.ScopeReadWrite, cfg.Insecure, cfg.EnableHTTP2)
+	enableHTTP2 := cfg.EnableHTTP2
+	if !hedging {
+		// enforce http2 for non-hedging bucket (PUT/LIST/DELETE)
```
Comment (dannykopping) [inline pkg/storage/chunk/gcp/gcs_object_client.go:89]: I'm a little nervous about this; there's nothing explicit disabling this for non-`GET` requests right now - except that we only hedge `GET` requests (right now).

Don't let this block you, but I'd hope we could make this a little stronger. This is how nasty bugs slip in IMHO.

File: pkg/storage/chunk/gcp/instrumentation.go [hunk 2]
```diff
@@ -52,13 +53,24 @@ func bigtableInstrumentation() ([]grpc.UnaryClientInterceptor, []grpc.StreamClie
 
 func gcsInstrumentation(ctx context.Context, scope string, insecure bool, http2 bool) (*http.Client, error) {
 	// start with default transport
```
Comment (dannykopping) [inline pkg/storage/chunk/gcp/instrumentation.go:55]: Is this comment now stale?

Can we add a comment here why we need to do this? You spent a long time debugging this, and I worry that if someone reverts this later they might re-introduce the weird bug you spent ages debugging.

**Review by dannykopping — APPROVED** [review]
Two non-blocking comments, but LGTM

File: pkg/storage/chunk/gcp/gcs_object_client.go (hunk 1 above)
Comment (cyriltovena) [reply pkg/storage/chunk/gcp/gcs_object_client.go:89]: yeah I can find a way to refactor this to make more clear. I went for the easier route.

Comment (slim-bean) [reply pkg/storage/chunk/gcp/gcs_object_client.go:89]: I'm not totally sure why we need this? the default is for hedging to be disabled and http2 to be enabled.

I'm not sure what scenario this protects against, what would be the problem of disabling http2 if someone wanted to and they were also not using hedging?

**Review by slim-bean — APPROVED** [review]
Also had a question on the same code Danny commented on, but otherwise LGTM!

File: pkg/storage/chunk/gcp/gcs_object_client.go (hunk 1 above)
Comment (owen-d) [reply pkg/storage/chunk/gcp/gcs_object_client.go:89]: Shouldn't hedging and http2 be independent options?

**Review by owen-d — COMMENTED** [review]
I'm a bit confused by this PR, but it's been a long day so I'll take another look tomorrow.

File: pkg/storage/chunk/gcp/gcs_object_client.go (hunk 1 above)
Comment (cyriltovena) [reply pkg/storage/chunk/gcp/gcs_object_client.go:89]: I'm trying to disable http2 only for Gets the code is clunky but does it. I'll refactor it to make it clear

File: pkg/storage/chunk/gcp/instrumentation.go [hunk 3]
```diff
@@ -52,13 +53,24 @@ func bigtableInstrumentation() ([]grpc.UnaryClientInterceptor, []grpc.StreamClie
...
+		MaxIdleConnsPerHost:   200,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+	}
 	if !http2 {
 		// disable HTTP/2 by setting TLSNextProto to non-nil empty map, as per the net/http documentation.
 		// see http2 section of https://pkg.go.dev/net/http
 		customTransport.TLSNextProto = make(map[string]func(string, *tls.Conn) http.RoundTripper)
+		customTransport.ForceAttemptHTTP2 = false
```
[the remark this answers is not in the document]
Comment (cyriltovena) [reply pkg/storage/chunk/gcp/instrumentation.go:72]: The default transport has it to `true` by default. I don't think it's required but added security we won't attempt http2 when we don;t want to.

**Review by dannykopping — APPROVED** [review]
LGTM, much cleaner - thanks :ok_hand:

**Pull request merged**